### ISSUE TITLE
Support range deletions in CQL

### ISSUE THREAD

**Issue by tzach** [scylladb/scylla#432]
from https://issues.apache.org/jira/browse/CASSANDRA-6237

given a table like:

```
CREATE TABLE events (
    id text,
    created_at timestamp,
    content text,
    PRIMARY KEY (id, created_at)
)
```

allow queries like:

```
DELETE FROM events WHERE id='someEvent' AND created_at < 'Jan 3, 2013';
```

**Comment by duarten** [comment]
I think this should also be marked for Cassandra 3.x compatability.

**Comment by duarten** [comment]
I see that @psarna was assigned to this, so a few words about the current state of range deletions.

TL;DR: we don't support infinite-bound range deletions or non-inclusive bounds in `compact storage` tables; we need support for the Cassandra 3.x sstable format before solving this.

Currently, we enforce that both range bounds in a range delete be specified, because we can't represent infinite bounds in the current sstable format. Such bounds are represented as a prefix with no components, with the `bound_kind` informing whether they are a bottom or top bound.  

We're currently unable to serialize an infinite bound in such a way that it would be correctly interpreted by Cassandra 2.2.x. A serialized bound is a composite with a `(<length><value><EOC>)+` format. While we could technically represent the bottom bound, the top bound would always sort before other values if written as a single component with 0 bytes in size and some `EOC`. The same would happen if represented as an empty (no components) composite, because in Cassandra 2.2.x those always have `EOC = NONE`.

We also cannot represent non-inclusive bounds on our current storage format for schemas that are non-compound, since the clustering key won't include the `EOC` byte.

**Comment by avikivity** [comment]
Let's decouple sstable capabilities from this.

`parsed_statement::prepare()` should receive some bools saying what we can and can't do (perhaps via an existing parameter, perhaps a new one).

Those parameters would then be updated by gossip. When gossip learns that the cluster supports the new format, those parameters are `true`d, and prepares start working.

@duarten would you list those parameters? only now-false parameters need be listed.

**Comment by tzach** [comment]
depended on #1969

**Comment by bhalevy** [comment]
With mc-format support what's left is support of open-ended range tombstones.

**Comment by tzach** [comment]
@psarna can we close this?

**Comment by tzach** [comment]
@psarna ping

**Comment by psarna** [comment]
@tzach sorry, missed that one. The short version is - I don't know.

There was an attempt to touch this issue in January, but as far as I can remember infinite bounds were still not 100% supported by then and I received no update regarding that topic. @bhalevy do you have any idea if  17acedb0c610d69cbc9675a02af467e44fa5b175 is still blocked by database capabilities (I think it was sstables 3.0 related?)?

Also, @eliransin, how does that issue fit with current priorities?

**Comment by bhalevy** [comment]
@psarna I tweaked your patch a bit, rebased, and fixed the unit test.
It now passes `test_range_deletion_scenarios`.
Please review PR #4591

**Comment by avikivity** [comment]
Not a bug, removing from bbackport candidates.
